tfjs-backend-webgpu/src/kernels/conv_backprop_mm_webgpu.ts

```diff
@@ -127,4 +131,70 @@ export class Conv2DDerInputMMProgram implements WebGPUProgram {
   `;
     return userCode;
   }
+
+  getUserCodeWgsl(): string {
+    const matMulSource =
+        makeMatMulPackedSourceWgsl(this.elementsPerThread, this.workGroupSize);
+
+    const readASnippet = `
+    let outRow = row / uniforms.outShape[2];
+    let outCol = row % uniforms.outShape[2];
+
+    let WRow = col / (uniforms.filterDims[1] * uniforms.outBackprop[3]);
+    let WCol = (col / uniforms.outBackprop[3]) % uniforms.filterDims[1];
+    let xR = f32(outRow - uniforms.pads[0] + WRow) / f32(uniforms.stride[0]);
+    let xC = f32(outCol - uniforms.pads[1] + WCol) / f32(uniforms.stride[1]);
+    if (xR < 0.0 || xR >= f32(uniforms.outBackprop[1]) || fract(xR) > 0.0) {
+      return 0.0;
+    }
+    if (xC < 0.0 || xC >= f32(uniforms.outBackprop[2]) || fract(xC) > 0.0) {
+      return 0.0;
+    }
+    let coord = vec4<u32>(
+        batch,
+        u32(xR),
+        u32(xC),
+        col % uniforms.outBackprop[3]);
+    return x.numbers[getFlatIndex4D(coord, uniforms.xShape)];`;
+
+    const sampleA = `if (row < uniforms.dimAOuter && col < uniforms.dimInner) {
+      ${readASnippet}
+    }
+    return 0.0;`;
+
+    const userCode = `
+    fn mm_readA(row : u32, col : u32, globalId : vec3<u32>) -> f32 {
+      var batch = globalId.z;
+      ${sampleA}
+    }
+
+    fn mm_readB(row : u32, col : u32, globalId : vec3<u32>) -> f32 {
+      let coordX = i32(uniforms.filterDims.x - 1u -
+          row / (uniforms.filterDims[1] * uniforms.outBackprop[3]));
+      let coordY = i32(uniforms.filterDims.y - 1u -
+          (row / uniforms.outBackprop[3]) % uniforms.filterDims[1]);
+      if (row < uniforms.dimInner && col < uniforms.dimBOuter &&
+          coordX >= 0 && coordY >= 0) {
+        let coord = vec4<u32>(u32(coordX), u32(coordY), col,
+            row % uniforms.outBackprop[3]);
+        return W.numbers[getFlatIndex4D(coord, uniforms.wShape)];
```

Comment:
This is almost the same as the glsl version, except >0 check.
BTW, I think we'd add coordsInBounds here, and also for glsl.

Reply:
For glsl, `if (row < dimInner && col < dimBOuter)` did the bounds checking. And I think >0 check here is not necessary here since there is not padding for filter.

Reply:
Sorry, I missed this one `if (row < uniforms.dimInner && col < uniforms.dimBOuter &&
          coordX >= 0 && coordY >= 0)`.